src/main/java/io/temporal/internal/statemachines/ActivityStateMachine.java

```diff
@@ -0,0 +1,338 @@
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License"). You may not
+ *  use this file except in compliance with the License. A copy of the License is
+ *  located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ *  or in the "license" file accompanying this file. This file is distributed on
+ *  an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ *  express or implied. See the License for the specific language governing
+ *  permissions and limitations under the License.
+ */
+
+package io.temporal.internal.statemachines;
+
+import static io.temporal.failure.FailureConverter.JAVA_SDK;
+
+import io.temporal.activity.ActivityCancellationType;
+import io.temporal.api.command.v1.Command;
+import io.temporal.api.command.v1.RequestCancelActivityTaskCommandAttributes;
+import io.temporal.api.command.v1.ScheduleActivityTaskCommandAttributes;
+import io.temporal.api.common.v1.Payloads;
+import io.temporal.api.enums.v1.CommandType;
+import io.temporal.api.enums.v1.EventType;
+import io.temporal.api.failure.v1.ActivityFailureInfo;
+import io.temporal.api.failure.v1.CanceledFailureInfo;
+import io.temporal.api.failure.v1.Failure;
+import io.temporal.api.history.v1.ActivityTaskCanceledEventAttributes;
+import io.temporal.api.history.v1.ActivityTaskCompletedEventAttributes;
+import io.temporal.api.history.v1.ActivityTaskFailedEventAttributes;
+import io.temporal.api.history.v1.ActivityTaskTimedOutEventAttributes;
+import io.temporal.internal.replay.ExecuteActivityParameters;
+import io.temporal.workflow.Functions;
+import java.util.Optional;
+
+final class ActivityStateMachine
+    extends EntityStateMachineInitialCommand<
+        ActivityStateMachine.State, ActivityStateMachine.Action, ActivityStateMachine> {
+
+  enum Action {
+    SCHEDULE,
+    CANCEL
+  }
+
+  enum State {
+    CREATED,
+    SCHEDULE_COMMAND_CREATED,
+    SCHEDULED_EVENT_RECORDED,
+    STARTED,
+    COMPLETED,
+    FAILED,
+    TIMED_OUT,
+    CANCELED,
+    SCHEDULED_ACTIVITY_CANCEL_COMMAND_CREATED,
+    SCHEDULED_ACTIVITY_CANCEL_EVENT_RECORDED,
+    STARTED_ACTIVITY_CANCEL_COMMAND_CREATED,
+    STARTED_ACTIVITY_CANCEL_EVENT_RECORDED,
+    COMPLETED_CANCEL_REQUESTED,
+    FAILED_CANCEL_REQUESTED,
+    CANCELED_CANCEL_REQUESTED,
+  }
+
+  private static StateMachine<State, Action, ActivityStateMachine> newStateMachine() {
+    return StateMachine.<State, Action, ActivityStateMachine>newInstance(
+            "Activity",
+            State.CREATED,
+            State.COMPLETED,
+            State.FAILED,
+            State.TIMED_OUT,
+            State.CANCELED)
+        .add(
+            State.CREATED,
+            Action.SCHEDULE,
+            State.SCHEDULE_COMMAND_CREATED,
+            ActivityStateMachine::createScheduleActivityTaskCommand)
+        .add(
+            State.SCHEDULE_COMMAND_CREATED,
+            CommandType.COMMAND_TYPE_SCHEDULE_ACTIVITY_TASK,
+            State.SCHEDULE_COMMAND_CREATED)
```

Comment:
is this a typo / copy paste error (SCHEDULE_COMMAND_CREATED twice)? Maybe I'm interpreting this incorrectly? I assumed first line is current state, subsequent line is potential transition, second line is final state and last line is an additional callback to run

Reply:
This is practically saying that this event is not causing a state transition. We could introduce another event here, but it doesn't really bring any value as no action is executed.